Update Service Aggregation Methods to use SQL

NickLaMuro opened the pull request:
## Purpose

Displaying the number "Totals for Service VMs" on the `/service/explorer/:id` page is expensive at scale.  The following table rows:
- CPU
- Memory
- Disk Count
- Disk Space Allocated
- Disk Space Used
- Memory on Disk

Currently requiring load in each hardware record for each vm record, and sometimes each disk record for those hardware records, causing a huge `N*N(N)? + 1`™ when a service has a lot of VMs.
## Overview (aka: WTF is even going on here...)

This is going to be a long, almost blog-post like, explanation here, so hold on to your butts...
### How we currently aggregate these stats...

Taking a step back and looking a particular example from one of the table rows above, to calculate (aggregate) the number of CPUs for a given service currently, you would call `my_service.aggregate_all_vm_cpus` and get returned the number of CPUs that are used by ALL VMs that were created using this service.  What happens under the hood is the following:
- `my_service.all_vms`
  - This is a `virtual_has_many`, but basically is makes use of `ancestry` on services to return all of the VMs that are associated with this service , and any other service that is a descendant of this one (uses the `.direct_vms` method on each service and child service in `app/models/service.rb` to do this).
- `.map(&:hardware)`
  - In reality, we are using a `virtual_column` here, but because it isn't making use of any arel, we are effectively doing a map.
- `.inject(0) {|sum, hw| sum += hw.cpu_total_cores.to_i`

With anything having to do with `Disks`, we are also effectively adding another `.flat_map(&:disks)` following the `.map(&:hardware)` step, which then totals up the values from the disk records.  As you probably can probably guess, this is extremely expensive with a service with many VMs.
### Representing the same thing with SQL

To represent the `aggregate_all_vm_cpus` as a column in SQL, you can represent `Service.select(:id, :aggregate_all_vm_cpus).first` as the following:

``` SQL
SELECT id,
      (
       --- Return only the SUM of all of the CPU cores to the main query (`.inject(0) { ... }`)
       SELECT SUM(hardwares.cpu_total_cores)

       --- Make sure we alias the table name so it doesn't conflict with the outer one
       FROM services aggregate_all_vm_cpus_services

       --- Join on services_resources that have a type of 'VmOrTemplate' (part of `.direct_vms`)
       JOIN service_resources ON service_resources.service_id = aggregate_all_vm_cpus_services.id
                             AND service_resources.resource_type = 'VmOrTemplate' 

       --- Include vms and their associated hardware (`.map(&:hardware)`)
       JOIN vms               ON vms.id = service_resources.resource_id
       JOIN hardwares         ON hardwares.vm_or_template_id = vms.id

       --- Include services that are the current service, and descendants of that service (`.all_vms`)
       WHERE (("aggregate_all_vm_cpus_services"."id" = services.id 
               OR "aggregate_all_vm_cpus_services"."ancestry" ILIKE CONCAT(services.id, '/%'))
              OR "aggregate_all_vm_cpus_services"."ancestry" = CAST(services.id AS VARCHAR))) AS aggregate_all_vm_cpus
FROM services
WHERE id = <my_service.id>;
```

This is not ideal, and should not be used when against a index action, as this nested `SELECT` would be run individually for each service, but for a show, it is about a 20ms nested select in the query for a service with 1800 VMs associated with it and it's children.

To do this for `Disk Count`, another join would need to be done for disks against the `hardware` records.
### Moving this into `virtual_attributes`

Part the reasoning behind doing this as a nested `SELECT` statement is it allows it to be used as a `virtual_attribute`/`virtual_column`.  This allows it to be queried against and selected as if it were a column on the `services` table:

``` ruby
irb> Service.select(:id, :aggregate_all_vm_cpus).where(:id => 123).first.attributes["aggregate_all_vm_cpus"]
=> 12345
```

And multiple of these can also then be used together in a single query:

``` ruby
irb> Service.select(:id, :aggregate_all_vm_cpus, :aggregate_all_vm_disk_count).where(:id => 123).first.attributes
  Service Load (146.1ms)  SELECT  "services"."id", ((SELECT ...
=> {"id" => 123, "aggregate_all_vm_cpus" => 12345, "aggregate_all_vm_disk_count" => 4321}
```

Which now takes a fraction of a second instead of multiple seconds when grabbing each record individually.  This did require modifying the underlying methods in order to check for the existence of the attribute before aggregating in ruby.
### Working it into `/service/explorer/:id`

For any of these new `virtual_attribute` arel methods to actually have an affect, they need to be included in the select in some way, otherwise they will just default to the original `N+1` method of aggregating the data.  This is done when calling `get_node_info` in the `if params[:id]` of the `explorer` action, which then calls `show_record`, then `identify_service`, then `identify_record` (defined in `ApplicationController::CiProcessing`), and finally calling `find_by_id_filtered` (defined in the main `ApplicationController`.

For simplicity, the `with_aggregates` scope was created to make including the columns easier.  This needed to be added to the query that returned the record that was acted against in the views so the aggregates aren't recalculated, so modifying `find_by_id_filtered` was necessary to prevent this.
## Metrics

The following metrics should be a clear indication of the improvements provided by this change.

**Before**

| ms | queries | query (ms) | rows |
| --: | --: | --: | --: |
| 32,534.7 | 278 | 2,922.5 | 74,540 |

**After**

| ms | queries | query (ms) | rows |
| --: | --: | --: | --: |
| 15,958.3 | 44 | 1,304.6 | 36,626 |
## Considerations and Mentionables
- Truth be told, we are technically moving the N+1 here from the application into the database, it just so happens that the DB is much more efficient at running this kind of stuff than the application is, and doesn't have nearly the object overhead that comes with running this in the DB.  That said, there will be probably a bit more stress put on the DB to execute these queries, but nothing in the sense of "multi second table locking", and overall it should be less of an impact on resources.
- These are DEFINITELY not queries that you want to run in a index type fashion, as they will not scale. This is because the nested selects will individually run for each service record returned.  In this case, it will end up in a tying up a process, so these methods need to be used with care.
- Optimally, sharing the `JOIN`s here would be ideal as they would prevent making multiple loops in the database for each aggregate column.  I was originally under the impression that the current UI doesn't facilitate for something like that easily, but it might be possible based on how I have implemented the scopes for this and the changes to the controller.  This would be a minor improvement at this point, and the `virtual_column` arel improvements might still want to be included regardless.
- The overwriting/modifications of `find_by_id_filtered` is ugly at best.  Unfortunately, we are inheriting a lot from the ApplicationController when finding records for "show" routes.  I was mostly going for allowing for additions (scopes on the finder), and overriding as few methods as possible when doing so.  `find_by_id_filtered` should still be the method that is used though at the end of the day and make sure we are finding only what the user has access to, and splitting it up would cause more chances for this filtering to be done incorrectly.
  - Something I noticed while working on that method is it makes two calls to the same record.  One for the `find_by`, and the other when tossed into `Rbac.filtered`.  Furthermore, this also blows away any scopes that might have been applied in the first finder `find_by` when tossed into Rbac, meaning I had to apply the `:named_scope` to Rbac or it wouldn't take.  Since normally this is used with just a model's class and an ID passed in, this is fine, but I found it confusing myself when trying to apply the scope originally to the first query.

NickLaMuro commented:
@miq-bot add_label performance, ui, refactoring

kbrock commented on app/models/service/aggregation.rb:
```diff
@@ -42,26 +151,36 @@ def aggregate_direct_vm_memory_on_disk
 
   def aggregate_all_vm_cpus
-    all_vms.inject(0) { |aggregate, vm| aggregate += vm.cpu_total_cores.to_i }
+    attributes["aggregate_all_vm_cpus"] || all_vms.inject(0) { |aggregate, vm| aggregate += vm.cpu_total_cores.to_i }
```
This runs into problems if the attribute is not defined.

please:

```ruby
if has_attribute?("aggregate_all_vm_cpus")
  attributes["aggregate_all_vm_cpus"]
else
  all_vms.inject(0) { |aggregate, vm| aggregate += vm.cpu_total_cores.to_i }
end
```

imtayadeway replied:
Minor, but will
```ruby
all_vms.inject(0) { |aggregate, vm| aggregate + vm.cpu_total_cores.to_i }
```
behave the same, avoiding the need to reassign here and elsewhere?

NickLaMuro replied:
> This runs into problems if the attribute is not defined.

@kbrock Not sure I follow the rational you have provided.  `attributes` is a hash, so if the hash key look up doesn't find a key, then it will continue down to the second part of the conditional and execute what was there originally.

That said, if you were arguing that this "stylistically is crap" or "not following conventions", then I would be fine on changing it for that reason alone.  Mostly did it this way to make the diff easier to grok as it only adds a check for a pre-calculated value first before moving on to the code that was there previously.

> Minor, but will `all_vms.inject(0) { |aggregate, vm| aggregate + vm.cpu_total_cores.to_i }` behave the same...

Yes, it should as far as I can tell, since the return value of `inject` should be the new value for the `aggregate`.  That said, I would argue if I am to do it, I should do it in a separate refactoring PR quickly first so I can touch the other aggregate methods in this PR that aren't part of this one.  Want me to do that?

kbrock replied:
Not talking style

For me, using `self["aggregate_all_vm_cpus"]` when the column is not in the query, it blows up for me.
If you use `has_attribute?("aggregate_all_vm_cpus")` and the column is not available, it works.

not intending on talking style

kbrock replied:
aaah - got it.

`attributes` creates a new attribute hash ever call. A bit of duplication.
Yes, what you said is true, it will silently fail for an unknown attribute and continue through.

I tend to use `self[attr]` which will not create that hash. And that version will blow up if it doesn't have the attribute.

NickLaMuro replied:
Gotcha.  If it is making a hash everytime, then yeah, maybe the other route is preferred.  Will investigate.

imtayadeway replied:
@NickLaMuro that makes sense, yes. But I'd say only if you felt especially moved to do so :wink:

NickLaMuro replied:
@imtayadeway Won't bother me as long as you are willing to 👍 it ;)

imtayadeway replied:
@NickLaMuro willing and able!

kbrock commented on app/controllers/application_controller.rb:
```diff
@@ -2279,7 +2279,7 @@ def find_by_id_filtered(db, id)
     end
 
-    Rbac.filtered_object(db_obj, :user => current_user) ||
+    Rbac.filtered_object(db_obj, :user => current_user, :named_scope => options[:rbac_scope]) ||
```
Could we use `:named_scope` coming into this method as well?
Just throwing it out there so we can keep parameters consistent

NickLaMuro replied:
Yeah, I am on the fence with this one.  On one hand, I agree from where you are coming from, but the other is me kind of making a subtle point about the fact that record that we are actually returning here is from the `Rbac` call, and not the `find_by(:id => 12345)` part.  This had me stumbling for about half a day digging into `Rbac#search` to determine if it made a second `SQL` call to determine if the record is accessible, and tried to make note of that (poorly) in the description:

> - Something I noticed while working on that method is it makes two calls to the same record.  One for the `find_by`, and the other when tossed into `Rbac.filtered`.  Furthermore, this also blows away any scopes that might have been applied in the first finder `find_by` when tossed into Rbac, meaning I had to apply the `:named_scope` to Rbac or it wouldn't take.  Since normally this is used with just a model's class and an ID passed in, this is fine, but I found it confusing myself when trying to apply the scope originally to the first query.

I also didn't want to overload `:named_scope` since it was passed around in other places in the controllers (probably mostly to pass into Rbac anyway), and this seemed like it would make it easier to grep for in the future.

Anyway, not tied to the name, just providing rational.  Let me know what you would prefer.  I have to rebase anyway, so making the change isn't a big deal.

kbrock replied:
think I'm running this `find_by` issue elsewhere too.

kbrock commented on app/controllers/service_controller.rb:
```diff
@@ -77,6 +77,7 @@ def explorer
     build_accordions_and_trees
 
     if params[:id]  # If a tree node id came in, show in one of the trees
+      @find_with_aggregates = true
```
heh... no comment

kbrock commented on app/models/service/aggregation.rb:
```diff
@@ -9,12 +9,121 @@ module Service::Aggregation
+      # aggregation_sql = aggregation_sql * 1.megabyte if options[:mb_to_bytes]
+      # aggregation_sql = Arel::Nodes::NamedFunction.new('NUMERIC', [aggregation_sql.sum])
+
+      aggregation_sql = Hardware.arel_table[hardware_column].sum
```
Is it possible to move these to delegates?

kbrock commented on app/models/service/aggregation.rb:
```diff
@@ -9,12 +9,121 @@ module Service::Aggregation
+        where_clause   = t.grouping(
+          subtree_services[:id].eq(t[:id])
+           .or(subtree_services[:ancestry].matches(ancestry_ilike))
+        ).or(subtree_services[:ancestry].eq(service_id_cast))
```
Think ancestry has all these defined as accessors. Can we use those?

NickLaMuro replied:
Looked into it and it seemed like it was calling `self.id`, meaning it was assuming that it was acting on an existing record downloaded from the DB, not the `id` column from the database.  I will double check this, since I would prefer to do this and re-use the scopes from the gem, but wrote it off as not possible.

kbrock commented on app/models/service/aggregation.rb:
```diff
@@ -9,12 +9,121 @@ module Service::Aggregation
+        end
+
+        subselect =
+          subtree_services.project(aggregation_sql)
```
I know I'm delegate happy. but when I see a subselect and deriving a join to the current table, I tend to think delegate

NickLaMuro replied:
So that is kind of what I was getting at (poorly) with this (again, quoting PR description):

> * Optimally, sharing the `JOIN`s here would be ideal as they would prevent making multiple loops in the database for each aggregate column. I was originally under the impression that the current UI doesn't facilitate for something like that easily, but it might be possible based on how I have implemented the scopes for this and the changes to the controller. This would be a minor improvement at this point, and the `virtual_column` arel improvements might still want to be included regardless.

Essentially, we could have the SQL be something more like this instead:

```SQL
SELECT id,
       SUM(hardwares.cpu_total_cores) AS aggregate_all_vm_cpus
       SUM(hardwares.memory_mb) AS aggregate_all_vm_memory
FROM services
JOIN service_resources ON service_resources.service_id = services.id
                       AND service_resources.resource_type = 'VmOrTemplate' 
JOIN vms               ON vms.id = service_resources.resource_id
JOIN hardwares         ON hardwares.vm_or_template_id = vms.id
WHERE (("services"."id" = services.id 
               OR "services"."ancestry" ILIKE CONCAT(services.id, '/%'))
              OR "services"."ancestry" = CAST(services.id AS VARCHAR)))
WHERE id = <my_service.id>;
```

And just have the scope `with_aggregates` include those columns and joins.  When I was originally working on this, I was under the impression that using the `virtual_attribute` pattern was going be required as this was pulled from a `MiqReport`, and only around Friday did I realize the record actually displaying these aggregates was not going through the report.  I had all this code worked up, so I wanted to put it up for review as is, but it might be that this isn't the best way to go about it.

"Hashtage #PipeDream":  Given the time and bandwidth to do this completely differently, I would argue that pulling in the stats as part of the main request is something we should not do, and we rub some of the "MSA Special Sauce" (Mirco Services Architecture) on this and make a separate request to retrieve these aggregate stats, and avoid slowing down the main query.

kbrock commented on app/models/service/aggregation.rb:
```diff
@@ -42,26 +151,36 @@ def aggregate_direct_vm_memory_on_disk
 
   def aggregate_all_vm_disk_count
-    all_vms.inject(0) { |aggregate, vm| aggregate += vm.num_disks.to_i }
+    attributes["aggregate_all_vm_disk_count"] || all_vms.inject(0) { |aggregate, vm| aggregate += vm.num_disks.to_i }
```
really like adding this pattern. would like to see the `has_attribute?` version though

NickLaMuro replied:
Discussed I think in https://github.com/ManageIQ/manageiq/pull/11502#discussion_r80530332, but I can switch to that if you would prefer.

kbrock commented:
I had added #11300 to help delegating to methods like these

kbrock commented on app/models/service/aggregation.rb:
```diff
@@ -9,12 +9,121 @@ module Service::Aggregation
-    virtual_column :aggregate_all_vm_disk_space_allocated,    :type => :integer
-    virtual_column :aggregate_all_vm_disk_space_used,         :type => :integer
-    virtual_column :aggregate_all_vm_memory_on_disk,          :type => :integer
+    virtual_attribute :aggregate_all_vm_cpus,   :integer,
```
I guess I like it better if we could define something like this:

```ruby
  virtual_aggregate :aggregate_all_vm_cpus, :cpus, :cpu_total_cores
```

kbrock replied:
see #11514

NickLaMuro commented:
Updated the code to fix cops.  Nothing has changed with the metrics, just breaking up the main method so `rubocop` is satisfied.

kbrock commented on app/models/service/aggregation.rb:
```diff
@@ -9,12 +9,115 @@ module Service::Aggregation
+
+    VM_OR_TEMPLATE_TYPE   = Arel::Nodes::SqlLiteral.new("'VmOrTemplate'")
+    ANCESTRY_MATCH        = Arel::Nodes::SqlLiteral.new("'/%'")
+    ANCESTRY_ILIKE        = Arel::Nodes::NamedFunction.new("CONCAT", [Service.arel_table[:id], ANCESTRY_MATCH])
```
Lets avoid having `arel_table` in constants.
Simply loading this module will force us to hit the db for `Service`

please put these into a lambda or method that is not called at boot time.

NickLaMuro replied:
I guess I am confused about the concern here.

As this is a database backed application, and making 1 extra request to the DB to initialize this model (for each constant the has `arel_table` in it of course) doesn't seem like it would have any huge impact that we should be concerned about (I realize that latency can be a factor when an appliance is setup against a remote DB, but if it is setup where latency is so bad that it can't perform a simple table `EXPLAIN` to build the `arel_table` object on application boot, my it will fall flat when at runtime when even some of the more optimized pages... say, what I changed in this PR).  

And to me, this **IS** a constant, and shouldn't be put into a method where it isn't rebuilt and a new object is instantiated each time the `virtual_attributes` are called, in which it would generate the same exact SQL each time.  For the `/service/explorer/:id`, when `.with_aggregates` is called, that means we are instantiating `ANCESTRY_ILIKE` (well, the method/variable version of it if it were changed) 6 times for each `virtual_attribute`.

I am fine if there is idle memory concerns, stale connection issues, or the `arel_table` will change over time and cause issues (with the use of `table_alias` or something), then I would be fine changing it, but if the concern is only loading it on boot then I am not sure I follow.

NickLaMuro replied:
Fixed.  Had a discussion about this off line and switched to memoized class methods to avoid the DB calls on class load.

Reason for the concern raised was that it can break building the appliance, since the DB is not always present in different phases of the build (or that is how I understood it).

NickLaMuro commented:
@miq-bot remove_label wip

NickLaMuro commented:
@Fryguy please review when you have a chance.

Fryguy commented:
I have a lot of reading to catch up on this one.  On the VERY casual review I did without actually reading anything, I will say the following...please take it with a grain of salt and I'll give a much deeper reading on Monday (hopefully 😛 )

First, I don't like all the Arel...I find if we need to drop down that low, I'm wondering if we are doing something wrong.  Perhaps *I'm* wrong.  I find when we drop down to Arel the code gets extremely complex to read, and generally there is a simpler way to bring the data back to Ruby efficiently and do the calculation there.  Yeah, it's a *little* slower or uses a little memory, but the maintenance value is MUCH easier.  For example, trying to get things into subselects with groupings and aliases, when a simple `.includes(:hardwares).each { |x| calculate }` will do.  Not sure...again I didn't read everything here (it's going to take a while 😄 )

Now, as far as aggregates go, it took me a while to figure out that you *aren't* talking about AggregationMixin.  In there we do optimized aggregation, but it doesn't look nearly as complicated as it does here.  Perhaps that's because of the ancestry layer, but I'm not sure yet.  I don't know why Service::Aggregation was built the way it was when we have AggregationMixin.  Have you reviewed AggregationMixin to see if it could be reused or if the techniques in there could be reused?

kbrock commented:
@Fryguy this looks good to me - anything on your side.
Maybe @chrisarcand has an opinion?

NickLaMuro commented:
> Yeah, it's a little slower or uses a little memory, ...

Working this in a more formal "report", but for now, the before and after for memory is about `300MB -> ~6MB` with this change.

NickLaMuro commented:
@Fryguy Responding to the rest of your comments:

> I don't know why Service::Aggregation was built the way it was when we have AggregationMixin. Have you reviewed AggregationMixin to see if it could be reused or if the techniques in there could be reused?

I have no idea why we didn't use that in for `Service`, but I did notice that it existed after Keenan pointed it out.  I don't know if using that would be useful, as it seems like some of the columns in there are different from the ones aggregated in `Service::Aggregation`, and I am not sure if that would mess with the other models that include that module.  I can do a spike in switching to that to see if it would be of some use.


> For example, trying to get things into subselects with groupings and aliases, when a simple .includes(:hardwares).each { |x| calculate } will do.

Not sure it is that simple.  To start, we are doing aggregations from the `service` level, meaning we are accessing VMs through the [`all_vms`](https://github.com/ManageIQ/manageiq/blob/master/app/models/service.rb#L114-L116) call.  Since this is using `MiqPreloader`'s `preload_and_map` on the `subtree`, and `direct_vms` calls a `collect` on `:resource`, so this probably trashes any rails relationships that we have and allowing us to use `includes`, which is why I built it into the the arel that was generated.  From @kbrock 's comments, it should be noted that we can't use the scopes from ancestry either since they reference `self.id` in them.

Avoiding the `includes` also avoids instantiating all of these heavy objects just to calculate these numbers.  I did some light testing in with basically the following two statements:

```ruby
irb> Service.where(:id => 12345678).first.aggregate_all_vm_cpus
irb> Service.where(:id => 12345678).with_aggregates.first.aggregate_all_vm_cpus
```

Here are the results when running it through [`MemoryProfiler`](https://github.com/SamSaffron/memory_profiler):


### Before

```ruby
irb(main):001:0> report = StringIO.new; MemoryProfiler.report { Service.where(:id => 12345678).first.aggregate_all_vm_cpus }.pretty_print(report); report.rewind; puts report.readlines[0..1]
Total allocated: 276579199 bytes (2580575 objects)
Total retained:  25413864 bytes (85328 objects)
=> nil
irb(main):002:0> report = StringIO.new; MemoryProfiler.report { Service.where(:id => 12345678).first.aggregate_all_vm_cpus }.pretty_print(report); report.rewind; puts report.readlines[0..1]
Total allocated: 111997928 bytes (1264737 objects)
Total retained:  272 bytes (3 objects)
=> nil
irb(main):003:0> report = StringIO.new; MemoryProfiler.report { Service.where(:id => 12345678).first.aggregate_all_vm_cpus }.pretty_print(report); report.rewind; puts report.readlines[0..1]
Total allocated: 111997928 bytes (1264737 objects)
Total retained:  272 bytes (3 objects)
=> nil
irb(main):004:0> ActiveSupport::NumberHelper.number_to_human_size 276579199
=> "264 MB"
irb(main):005:0> ActiveSupport::NumberHelper.number_to_human_size 111997928
=> "107 MB"
```


### After

```rubyirb(main):001:0> report = StringIO.new; MemoryProfiler.report { Service.where(:id => 12345678).with_aggregates.first.aggregate_all_vm_cpus }.pretty_print(report); report.rewind; puts report.readlines[0..1]
Total allocated: 7160726 bytes (65234 objects)
Total retained:  1575598 bytes (8304 objects)
=> nil
irb(main):002:0> report = StringIO.new; MemoryProfiler.report { Service.where(:id => 12345678).with_aggregates.first.aggregate_all_vm_cpus }.pretty_print(report); report.rewind; puts report.readlines[0..1]
Total allocated: 104832 bytes (1554 objects)
Total retained:  0 bytes (0 objects)
=> nil
irb(main):003:0> report = StringIO.new; MemoryProfiler.report { Service.where(:id => 12345678).with_aggregates.first.aggregate_all_vm_cpus }.pretty_print(report); report.rewind; puts report.readlines[0..1]
Total allocated: 104494 bytes (1552 objects)
Total retained:  0 bytes (0 objects)
=> nil
irb(main):004:0> ActiveSupport::NumberHelper.number_to_human_size 7160726
=> "6.83 MB"
irb(main):005:0> ActiveSupport::NumberHelper.number_to_human_size 104832
=> "102 KB"
```

The changes in memory usage after the first on each of those can probably be attributed to the ActiveRecord query cache being enabled.  If we change out `aggregate_all_vm_cpus` for `aggregate_all_vm_disk_count`, which now requires loading the `disk` records for each `hardware` record, then we have the following:


### Before

```ruby
irb(main):001:0> report = StringIO.new; MemoryProfiler.report { Service.where(:id => 12345678).with_aggregates.first.aggregate_all_vm_disk_count }.pretty_print(report); report.rewind; puts report.readlines[0..1]
Total allocated: 7160686 bytes (65233 objects)
Total retained:  1575598 bytes (8304 objects)
=> nil
irb(main):002:0> report = StringIO.new; MemoryProfiler.report { Service.where(:id => 12345678).with_aggregates.first.aggregate_all_vm_disk_count }.pretty_print(report); report.rewind; puts report.readlines[0..1]
Total allocated: 104792 bytes (1553 objects)
Total retained:  0 bytes (0 objects)
=> nil
irb(main):003:0> report = StringIO.new; MemoryProfiler.report { Service.where(:id => 12345678).with_aggregates.first.aggregate_all_vm_disk_count }.pretty_print(report); report.rewind; puts report.readlines[0..1]
Total allocated: 104454 bytes (1551 objects)
Total retained:  0 bytes (0 objects)
=> nil
irb(main):004:0> ActiveSupport::NumberHelper.number_to_human_size 7160686
=> "6.83 MB"
irb(main):005:0> ActiveSupport::NumberHelper.number_to_human_size 104454
=> "102 KB"
```


### After

```ruby
irb(main):001:0> report = StringIO.new; MemoryProfiler.report { Service.where(:id => 12345678).first.aggregate_all_vm_disk_count }.pretty_print(report); report.rewind; put
s report.readlines[0..1]
Total allocated: 375591683 bytes (3884256 objects)
Total retained:  25417360 bytes (85373 objects)
=> nil
irb(main):002:0> report = StringIO.new; MemoryProfiler.report { Service.where(:id => 12345678).first.aggregate_all_vm_disk_count }.pretty_print(report); report.rewind; puts report.readlines[0..1]
Total allocated: 211416769 bytes (2570130 objects)
Total retained:  272 bytes (3 objects)
=> nil
irb(main):003:0> report = StringIO.new; MemoryProfiler.report { Service.where(:id => 12345678).first.aggregate_all_vm_disk_count }.pretty_print(report); report.rewind; puts report.readlines[0..1]
Total allocated: 211416769 bytes (2570130 objects)
Total retained:  272 bytes (3 objects)
=> nil
irb(main):004:0> ActiveSupport::NumberHelper.number_to_human_size 375591683
=> "358 MB"
irb(main):005:0> ActiveSupport::NumberHelper.number_to_human_size 211416769
=> "202 MB"
```

I will let you be the judge if that is _"a little memory"_, but this is only a portion of the request that I am profiling here.

NickLaMuro commented:
Rebased with the following changes:

* Added documentation for new methods (in a separate commit for now)
* Removed unused option
* Removed unused join on partitions
* Removed need to pass around `aggregation_sql` variable to other methods

kbrock commented:
kicked the build. think it was unrelated.

Fryguy commented:
> I will let you be the judge if that is "a little memory", but this is only a portion of the request that I am profiling here

Sorry, I when I mentioned "a little memory" I was talking about the delta over alternate approaches.  By that I mean, given your change gets it from 300MB to 6MB, if there is an alternate approach that is easier to maintain, but only gets it from, say, 300MB to 10MB, then it might be better to use that one even though it's 4MB more (that's the "a little memory" I meant)

NickLaMuro commented:
@Fryguy So that is where I am not sure how you can do this in Ruby without resorting to some kind of SQL solution.

What exists now is an N+1 for fetching all of the objects, which is slow (strictly because of the individual SQL calls), and uses a lot of memory in the ruby process because we are instantiating the those objects with all of there column data to perform the data aggregations.  If we were to use `includes`, we remove the N+1, sure, but we still instantiate the same number of objects to build those includes and should, in theory, use the same amount of memory.

We can limit this possibly by reducing the number of columns we return with the includes, meaning there is less for the ruby process to interpret from the database (buildings strings on the database adapter), and less overall in the ActiveRecord objects themselves, but the default ActiveRecord weight of those objects probably won't get us below the 50MB mark when dealing with 1k+ vms per service.

The other problem with a `includes` solution in this case is `Service` is a polymorphic relationship with VMs, so it is tough to do includes from that level.  I am getting around it by hardcoding it in the aggregation arel, but I am unsure of a clean way to do it in ruby.

NickLaMuro commented:
I have also looked into the `AggregationMixin` a bit and it doesn't seem like it is a clean swap in and replace, and would also require some view changes to make this work in the front end so we are calling the proper methods names that differ between `AggregationMixin` and `ServiceAggregation`.

dmetzger57 commented:
@NickLaMuro @kbrock @Fryguy this  PR is getting a bit stale, I believe the open question / issue is if the time saving (32 seconds down to 16 seconds) and memory savings (300Mb to 6Mb) for the scale env tested are worth the complexity or not. I've not heard a simpler solution discussed that provides a comparable level of win, so it's either we go with the added complexity or move on and live with the current solution for now - unless I missed a viable alternative. Looking for closure on a non-trivial request. Thoughts?

kbrock commented:
No, `AggregationMixin` does not work. It just encourages N+1 / same thing we are doing. It is also going against other columns and not meant to go across resources.

The main problems (that this tackles)
1. sql to join to all children of a service. Ancestry assumes that every parent id will be loaded into ruby. Fails for a join. (this is a big problem)
2. The polymorphic service_resources are a little generic. This is not really a simple `:through`, so I'm not sure if my dream of using `virtual_aggregates` would even work here.

Nick: if you think we could tear this out if we got a better way of doing joins or sums, then I'm all for going this route.

NickLaMuro commented:
As far as being able to replace this in the future, I didn't really add much here that is needed in other places, so I don't why not, and this definitely something that I think should be done in the future since I think the overall pattern we are doing here is a bit flawed (kinda what I was getting at in "point 3" in the "Considerations" section in the description).  I see this as a better to be fetched as a "Query Object", but that was a bigger hammer that I didn't want to invest in when I put this together in the past.

Fryguy commented:
@dmetzger57 I agree.  Although I don't really like all of the Arel, this PR does solve the problem, and we've not come up with anything better, so on that note, I am :+1: on this PR.

I believe @NickLaMuro said he had to fix something in this PR due to another PR getting merge which creates a logical conflict (as opposed to a hard git conflict).

Additionally, I found a minor typo 😝  `base_service_aggrigation_join` should `base_service_aggregation_join`

NickLaMuro commented:
@Fryguy fixed, though, I might do some quick performance tests with these changes just to be sure.

NickLaMuro commented:
New Metrics:

**Before**

|       ms |  queries |  query (ms) |      rows |
|     ---: |     ---: |        ---: |      ---: |
|   37,057 |    7,673 |       3,018 |    74,323 |
|   40,147 |    7,486 |       2,724 |   102,413 |
|   40,239 |    7,486 |       2,701 |   102,413 |
|   41,698 |    7,486 |       2,701 |   102,413 |
|   41,160 |    7,486 |       2,688 |   102,413 |

**After**

|       ms | queries |  query (ms) |     rows |
|     ---: |    ---: |        ---: |     ---: |
|   18,691 |     277 |       2160 |   36,409 |
|   20,475 |      91 |       1227 |   64,499 |
|   21,338 |      91 |       1083 |   64,499 |
|   21,230 |      91 |       1096 |   64,499 |
|   21,482 |      91 |       1090 |   64,499 |

imtayadeway commented on app/models/service/aggregation.rb:
```diff
@@ -9,12 +9,243 @@ module Service::Aggregation
+    end
+
+    # Generates a nested select statement that builds off of the `id` of the
+    # record being queried against(or multiple records, though this will not
```
Nit, but s/(/ (/

imtayadeway commented on app/models/service/aggregation.rb:
```diff
@@ -9,12 +9,243 @@ module Service::Aggregation
+    # @return [Arel::Nodes::And] The Arel conditional for JOINing the VMs table
+    def vm_join_clause(options = {})
+      clause = vms_tbl[:id].eq(service_resources_tbl[:resource_id])
+      if options[:on_vms_only] # meaning VMs that are powerd "on"
```
s/powerd/powered/

imtayadeway commented on app/models/service/aggregation.rb:
```diff
@@ -9,12 +9,243 @@ module Service::Aggregation
+    end
+
+    # NOTE: The following class methods are technically "constants", but to
+    # make sure they don't load with the class is first loaded and cause
```
Is this supposed to say "when the class"?

NickLaMuro replied:
Probably...

NickLaMuro commented on app/models/service/aggregation.rb:
```diff
@@ -42,26 +273,52 @@ def aggregate_direct_vm_memory_on_disk
-    all_vms.inject(0) { |aggregate, vm| aggregate + vm.ram_size.to_i }
+    if has_attribute?("aggregate_all_vm_memory")
+      # Avoids (poorly) "ERROR:  integer out of range" in postgres
+      self["aggregate_all_vm_memory"] * 1.megabyte
```
P.S.  Yes, I need to remove this from #13117 as for sure as well.

Fryguy commented on app/models/service/aggregation.rb:
```diff
@@ -42,26 +273,51 @@ def aggregate_direct_vm_memory_on_disk
-    all_vms.inject(0) { |aggregate, vm| aggregate + vm.ram_size_in_bytes_by_state.to_i }
+    if has_attribute?("aggregate_all_vm_memory_on_disk")
+      # Avoids (poorly) "ERROR:  integer out of range" in postgres
+      self["aggregate_all_vm_memory_on_disk"] * 1.megabyte
```
So in re-review, this comment concerns me.  This implies that the math is being done on ints instead of bigint.  Can we just cast to bigint?

NickLaMuro replied:
I want to say it wasn't as simple as doing that, and it was easier to do that one multiplication is ruby instead of trying to cast it in SQL (again, memory failing me here).

That said, it may have been as simple as "hmm... maybe I should make this as simple as I can for @Fryguy to approve this and not add more Arel...", which adding another cast would do.  Not really opinionated one way or the other at this point, but would need to re-test things if you want the change.

Fryguy commented on app/models/service/aggregation.rb:
```diff
@@ -9,12 +9,243 @@ module Service::Aggregation
+    end
+
+    def service_id_cast
+      @service_id_cast ||= Arel::Nodes::NamedFunction.new("CAST", [Service.arel_table[:id].as("VARCHAR")])
```
Why the cast to varchar?  An id is a bigint.  Is this make it work with ancestry strings?

NickLaMuro replied:
I wish you would have asked this 3 months ago when I did this so I had this fresh in my mind... I will have to go back over what I did and see why I did this, but I think you are right now that you mention it, I think this is because it is being used in comparisons for ancestry so those can be done in SQL.

NickLaMuro replied:
@Fryguy yes, you are correct since we are comparing against the ancestry column, since it is a string:

https://github.com/ManageIQ/manageiq/blob/master/db/migrate/20130923182042_collapsed_initial_migration.rb#L2583

https://github.com/NickLaMuro/manageiq/blob/99ff33aa/app/models/service/aggregation.rb#L153

Fryguy commented:
The failing test is due to a failure on master which is now resolved.  Merging.

chessbyte commented:
Helps performance of issues reported in https://bugzilla.redhat.com/show_bug.cgi?id=1403509

chessbyte commented:
Darga Backport conflict:
```sh
$ git cherry-pick -e -x -m 1 bbe42e0  
error: could not apply bbe42e0... Merge pull request #11502 from NickLaMuro/sqlize_service_aggregation_virtual_cols
hint: after resolving the conflicts, mark the corrected paths
hint: with 'git add <paths>' or 'git rm <paths>'
hint: and commit the result with 'git commit'

$ git status
On branch darga
Your branch is up-to-date with 'upstream/darga'.
You are currently cherry-picking commit bbe42e0.
  (fix conflicts and run "git cherry-pick --continue")
  (use "git cherry-pick --abort" to cancel the cherry-pick operation)

Changes to be committed:

	modified:   app/controllers/service_controller.rb

Unmerged paths:
  (use "git add <file>..." to mark resolution)

	both modified:   app/controllers/application_controller.rb
	both modified:   app/models/service/aggregation.rb

$ git diff
```

```diff
diff --cc app/controllers/application_controller.rb
index bd67b54,e240c0d..0000000
--- a/app/controllers/application_controller.rb
+++ b/app/controllers/application_controller.rb
@@@ -2351,15 -2204,16 +2351,19 @@@ class ApplicationController < ActionCon
    end
  
    # Following 3 methods moved here to ensure they are loaded at the right time and will be available to all controllers
-   def find_by_id_filtered(db, id)
+   def find_by_id_filtered(db, id, options = {})
      raise _("Invalid input") unless is_integer?(id)
  
 -    db_obj = db.find_by(:id => from_cid(id))
 -    if db_obj.nil?
 +    unless db.where(:id => from_cid(id)).exists?
        msg = _("Selected %{model_name} no longer exists") % {:model_name => ui_lookup(:model => db.to_s)}
        raise msg
      end
  
++<<<<<<< HEAD
 +    Rbac.filtered(db, :conditions => ["#{db.table_name}.id = ?", id], :user => current_user).first ||
++=======
+     Rbac.filtered_object(db_obj, :user => current_user, :named_scope => options[:named_scope]) ||
++>>>>>>> bbe42e0... Merge pull request #11502 from NickLaMuro/sqlize_service_aggregation_virtual_cols
        raise(_("User '%{user_id}' is not authorized to access '%{model}' record id '%{record_id}'") %
                {:user_id   => current_userid,
                 :record_id => id,
diff --cc app/models/service/aggregation.rb
index aa45b35,f9860c4..0000000
--- a/app/models/service/aggregation.rb
+++ b/app/models/service/aggregation.rb
@@@ -42,26 -273,51 +273,75 @@@ module Service::Aggregatio
    end
  
    def aggregate_all_vm_cpus
++<<<<<<< HEAD
 +    all_vms.inject(0) { |aggregate, vm| aggregate += vm.cpu_total_cores.to_i }
 +  end
 +
 +  def aggregate_all_vm_memory
 +    all_vms.inject(0) { |aggregate, vm| aggregate += vm.ram_size.to_i }
 +  end
 +
 +  def aggregate_all_vm_disk_count
 +    all_vms.inject(0) { |aggregate, vm| aggregate += vm.num_disks.to_i }
 +  end
 +
 +  def aggregate_all_vm_disk_space_allocated
 +    all_vms.inject(0) { |aggregate, vm| aggregate += vm.allocated_disk_storage.to_i }
 +  end
 +
 +  def aggregate_all_vm_disk_space_used
 +    all_vms.inject(0) { |aggregate, vm| aggregate += vm.used_disk_storage.to_i }
 +  end
 +
 +  def aggregate_all_vm_memory_on_disk
 +    all_vms.inject(0) { |aggregate, vm| aggregate += vm.ram_size_in_bytes_by_state.to_i }
++=======
+     if has_attribute?("aggregate_all_vm_cpus")
+       self["aggregate_all_vm_cpus"]
+     else
+       all_vms.inject(0) { |aggregate, vm| aggregate + vm.cpu_total_cores.to_i }
+     end
+   end
+ 
+   def aggregate_all_vm_memory
+     if has_attribute?("aggregate_all_vm_memory")
+       self["aggregate_all_vm_memory"]
+     else
+       all_vms.inject(0) { |aggregate, vm| aggregate + vm.ram_size.to_i }
+     end
+   end
+ 
+   def aggregate_all_vm_disk_count
+     if has_attribute?("aggregate_all_vm_disk_count")
+       self["aggregate_all_vm_disk_count"]
+     else
+       all_vms.inject(0) { |aggregate, vm| aggregate + vm.num_disks.to_i }
+     end
+   end
+ 
+   def aggregate_all_vm_disk_space_allocated
+     if has_attribute?("aggregate_all_vm_disk_space_allocated")
+       self["aggregate_all_vm_disk_space_allocated"]
+     else
+       all_vms.inject(0) { |aggregate, vm| aggregate + vm.allocated_disk_storage.to_i }
+     end
+   end
+ 
+   def aggregate_all_vm_disk_space_used
+     if has_attribute?("aggregate_all_vm_disk_space_used")
+       self["aggregate_all_vm_disk_space_used"]
+     else
+       all_vms.inject(0) { |aggregate, vm| aggregate + vm.used_disk_storage.to_i }
+     end
+   end
+ 
+   def aggregate_all_vm_memory_on_disk
+     if has_attribute?("aggregate_all_vm_memory_on_disk")
+       # Avoids (poorly) "ERROR:  integer out of range" in postgres
+       self["aggregate_all_vm_memory_on_disk"] * 1.megabyte
+     else
+       all_vms.inject(0) { |aggregate, vm| aggregate + vm.ram_size_in_bytes_by_state.to_i }
+     end
++>>>>>>> bbe42e0... Merge pull request #11502 from NickLaMuro/sqlize_service_aggregation_virtual_cols
    end
  end
```

NickLaMuro commented:
@chessbyte I don't think this should be backported to darga as it is taken care of in https://github.com/ManageIQ/manageiq/pull/13117

chessbyte commented:
Backported to Darga via #13117

simaishi commented:
Euwe backport details:

```
$ git log -1
commit ca34e404e4d911084da5e20e7c8d61fc49cf75cd
Author: Jason Frey <fryguy9@gmail.com>
Date:   Tue Dec 13 20:45:50 2016 -0500

    Merge pull request #11502 from NickLaMuro/sqlize_service_aggregation_virtual_cols
    
    Update Service Aggregation Methods to use SQL
    (cherry picked from commit bbe42e02017aa9a9a16c1e52a318daafb3a7a7b7)
    
    https://bugzilla.redhat.com/show_bug.cgi?id=1411883
    https://bugzilla.redhat.com/show_bug.cgi?id=1411887
```
